### ISSUE TITLE
Recent changes in handling of PMIX_EVENT_JOB_END?

### ISSUE THREAD

**Issue by drwootton** [openpmix/prrte#953]
Thank you for taking the time to submit an issue!

### What version of the PMIx Reference Server are you using? (e.g., v1.0, v2.1, git master @ hash, etc.)
Built from master source as of May 3

##### What version of PMIx are you using? (e.g., v1.2.5, v2.0.3, v2.1.0, git branch name and hash, etc.)
Built from master source as of May 3

### Please describe the system on which you are running

* Operating system/version:
RHEL 7.7

* Computer hardware:
Power 9 node 20 cores, 160 threads

* Network type:
localhost

-----------------------------

## Details of the problem

The indirect (indirect prterun -n 2 hello 30) and attach (attach &LT;prterun-namespace&GT;)examples were repeatedly running successfully last week (Wednesday) in my single node tests

I rebuilt from latest PMIx and PRRTE source Monday morning. Now the attach example hangs waiting for application termination and the indirect example has incorrect output. Other examples still run correctly.

The problem with the attach example appears to be that it no longer gets the PMIX_EVENT_JOB_END event from the prterun process it attached to or the daemon process it spawned. It does get a lost connection event from the prterun process, but that's not the event it was expecting.

In the indirect case, it looks like the PMIX_EVENT_JOB_END event from the prterun process gets sent to the indirect process instead and the indirect process is getting a PMIX_EVENT_JOB_END event from the daemon it spawned, both occuring after a lost connection event from the prterun process.

Did something change in the handling of these events in the past few days?

**Comment by rhc54** [comment]
Not that I'm aware of, no. I was a bit confused by your description's use of the term "daemon" vs "prterun", but I think I finally figured out what you meant. I can try to take a peek at it later in the week, or perhaps next week.

**Comment by drwootton** [comment]
Update, hopefully for clarification ...

I'm using prterun as the launcher for the application program in both cases, and the daemon is the PRRTE example daemon which reads the proc table for the application. In the attach case, I launch the application as **prterun -n 2 --report-uri + ./hello 30** and grab (copy) the URI reported by prterun. Then I start attach with the copied URI s the single command line parameter to attach. The attach program launches the daemon process. 

In the indirect case, I run **indirect prterun -n 2 ./hello **10** where indirect launches prterun then launches the daemon process that reads the proc table.

**Comment by rhc54** [comment]
FWIW: quick investigation indicates we have never sent a `PMIX_EVENT_JOB_END` event. The event is not present anywhere in PRRTE outside of your example codes. I'll have to add it - technically, we are supposed to generate it since you passed `PMIX_NOTIFY_JOB_EVENTS` in the spawn attributes.

However, it is also clear that your `indirect` example is not doing what you described above. There is no registered handler looking for the `PMIX_EVENT_JOB_END` event. Instead, you declare the IL to have terminated if _any_ event is received in the default event handler, or if the "lost connection" event is generated by the client library when it sees `prterun` terminate.

I can/will modify `indirect` so it is looking for the `PMIX_EVENT_JOB_END` of the application being executed (not `prterun`) as `prterun` cannot tell you of its own demise - that's what "lost connection" is for. I'll also modify PRRTE so we generate that event when requested.

I'll have to leave it to you to modify the logic in `indirect` so it terminates when it should.

**Comment by rhc54** [comment]
My bad - I realized today that we had renamed `PMIX_ERR_JOB_TERMINATED` to `PMIX_EVENT_JOB_END` in the standard, but not in the PRRTE code. I'll take care of that one so we avoid this confusion again.

So we have been generating that event. What puzzled me is that the `indirect` example doesn't seem to be looking for it. You set the variable `code` to that event when the variable is initially declared, but then you overwrite it with `PMIX_ERR_LOST_CONNECTION` before you ever use it. I therefore suspect the problem lies in a change to the `indirect` code and not PRRTE itself.

I'll verify that and comment back here.
